activation/certifier.go

```diff
@@ -0,0 +1,366 @@
+	Signature []byte `json:"signature"`
+	PubKey    []byte `json:"pub_key"`
+}
+
+type Certifier struct {
+	logger *zap.Logger
+	db     *localsql.Database
+	client certifierClient
+}
+
+type NewCertifierOption func(*Certifier)
+
+func WithCertificates(certs []Certificate) NewCertifierOption {
+	return func(c *Certifier) {
+		c.logger.Info("adding certificates", zap.Int("num", len(certs)), zap.Any("certs", certs))
+		for _, cert := range certs {
+			if err := certifier_db.AddCertificate(c.db, c.client.Id(), cert.Certificate.Inner, cert.Poet); err != nil {
+				c.logger.Warn("failed to persist poet cert", zap.Error(err))
+			}
+		}
+	}
+}
+
+func NewCertifier(
+	db *localsql.Database,
+	logger *zap.Logger,
+	client certifierClient,
+	opts ...NewCertifierOption,
+) *Certifier {
+	c := &Certifier{
+		client: client,
+		logger: logger,
+		db:     db,
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c
+}
+
+func (c *Certifier) GetCertificate(poet string) PoetCert {
+	cert, err := certifier_db.Certificate(c.db, c.client.Id(), poet)
+	switch {
+	case err == nil:
+		return cert
+	case !errors.Is(err, sql.ErrNotFound):
+		c.logger.Warn("failed to get certificate", zap.Error(err))
+	}
+	return nil
+}
+
+func (c *Certifier) Recertify(ctx context.Context, poet PoetClient) (PoetCert, error) {
+	url, pubkey, err := poet.CertifierInfo(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("querying certifier info: %w", err)
+	}
+	cert, err := c.client.Certify(ctx, url, pubkey)
+	if err != nil {
+		return nil, fmt.Errorf("certifying POST for %s at %v: %w", poet.Address(), url, err)
+	}
+
+	if err := certifier_db.AddCertificate(c.db, c.client.Id(), cert.Bytes(), poet.Address()); err != nil {
+		c.logger.Warn("failed to persist poet cert", zap.Error(err))
+	}
+
+	return cert, nil
+}
+
+// CertifyAll certifies the nodeID for all poets that require a certificate.
+// It optimizes the number of certification requests by taking a unique set of
+// certifiers among the given poets and sending a single request to each of them.
+// It returns a map of a poet address to a certificate for it.
+func (c *Certifier) CertifyAll(ctx context.Context, poets []PoetClient) map[string]PoetCert {
+	certs := make(map[string]PoetCert)
+	poetsToCertify := []PoetClient{}
+	for _, poet := range poets {
+		if cert := c.GetCertificate(poet.Address()); cert != nil {
+			certs[poet.Address()] = cert
+		} else {
+			poetsToCertify = append(poetsToCertify, poet)
+		}
+	}
+	if len(poetsToCertify) == 0 {
+		return certs
+	}
+
+	type certInfo struct {
+		url    *url.URL
+		pubkey []byte
+		poet   string
+	}
+
+	certifierInfos := iter.Map(poetsToCertify, func(p *PoetClient) *certInfo {
+		poet := *p
+		url, pubkey, err := poet.CertifierInfo(ctx)
+		if err != nil {
+			c.logger.Warn("failed to query for certifier info", zap.Error(err), zap.String("poet", poet.Address()))
+			return nil
+		}
+		return &certInfo{
+			url:    url,
+			pubkey: pubkey,
+			poet:   poet.Address(),
+		}
+	})
```

Comment:
Why the extra dependency?

```suggestion
	certifierInfos := make([]*certInfo, len(poetsToCertify))
	var eg errgroup.Group
	for i := range poetsToCertify {
		i := i
		eg.Go(func() err {
			url, pubkey, err := poetsToCertify[i].CertifierInfo(ctx)
			if err != nil {
				c.logger.Warn("failed to query for certifier info", zap.Error(err), zap.String("poet", poetsToCertify[i].Address()))
				return nil
			}
			certifierInfos[i] = &certInfo{
				url:    url,
				pubkey: pubkey,
				poet:   poetsToCertify[i].Address(),
			}
		})
	}
	eg.Wait()
```

Reply:
Alternatively I think it would also be good to have this function implemented once for one poet. Instead of iterating over all PoETs that went through a step (check if certificate is needed, fetch the CertifierInfo, fill the `certSvcs` map, Certify), it could instead call this one function for every PoET in an `errgroup` and return early when a PoET doesn't need to continue with the next step. Wdyt?

Reply:
I find the former version less cluttered, easier to read, and less error-prone.

>Alternatively I think it would also be good to have this function implemented once for one poet. Instead of iterating over all PoETs that went through a step (check if certificate is needed, fetch the CertifierInfo, fill the certSvcs map, Certify), it could instead call this one function for every PoET in an errgroup and return early when a PoET doesn't need to continue with the next step.

This code:
1. queries all poets for certifier info.
2. creates a unique set of certifiers to certify with
3. calls /certify on each

The point is to limit the number of calls on p. 3. All SM poets use the same certifier service and can reuse the same certificate so it makes sense to avoid duplicated calls to /certify.

Reply:
> I find the former version less cluttered, easier to read, and less error-prone.

It's literally the same code 🤷  just instead of

```go
certifierInfos := iter.Map(poetsToCertify, func(p *PoetClient) *certInfo {
    ...
})
```
it's
```
certifierInfos := make([]*certInfo, len(poetsToCertify))
var eg errgroup.Group
eg.Go(func() error {
    ... // (same as above)
})
eg.Wait()
```
That little bit extra code safes a whole extra dependency?

Reply:
Then it's not literally the same code... (btw you missed the for loop).

What's wrong with a dependency? We depend on it indirectly already so using it doesn't add any build time. IMHO, the https://pkg.go.dev/github.com/sourcegraph/conc#section-readme library contains lots of useful building blocks making concurrent Go easier, less cluttered, and safer (given its 7.6k stars it's not something "random"). We could adapt it in the code base in more places.

Reply:
I was referring to `...` as being the essentially the same code.

Generally I think we should aim to keep the amount of dependencies we use to a minimum. It might be the case that it is already a transient dependency, but those are (hopefully) managed by our direct dependencies. I don't think we need to start using a library for something that a) is simple to implement with the dependencies we already have and b) is arguably just as easy/hard to read:

The "self-made" approach is just using an `errgroup` in a for loop. In conctrast I had to check documentation / source code to see that `iter.Map` processes elements concurrently.

Reply:
>b) is arguably just as easy/hard to read
That's totally subjective. Also, one could say he needed to check `errgroup` documentation to see what it does (or any other library one is not familiar with yet) :shrug:. We already have 57 deps, some of which are doubtful (4 GH stars etc). In contrast, this is a well-known one with 7k+ stars.